modules/data/tools/smart_recorder/record_processor.h

```diff
@@ -0,0 +1,67 @@
+/******************************************************************************
+ * Copyright 2019 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#pragma once
+
+#include <memory>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "cyber/record/record_message.h"
+#include "cyber/record/record_writer.h"
+
+#include "modules/data/tools/smart_recorder/proto/smart_recorder_triggers.pb.h"
+#include "modules/data/tools/smart_recorder/trigger_base.h"
+
+namespace apollo {
+namespace data {
+
+using cyber::record::RecordMessage;
+using cyber::record::RecordWriter;
+
+/**
+ * @class RecordProcessor
+ * @brief Process messages and apply the rules based on configured triggers
+ */
+class RecordProcessor {
+ public:
+  RecordProcessor(std::string source_record_dir,
+                  std::string restored_output_dir);
+  bool Init(const SmartRecordTrigger& trigger_conf);
+  bool Process();
+  ~RecordProcessor() { writer_->Close(); }
+
+ private:
+  void LoadSourceRecords();
+  void CollectAllChannels(const SmartRecordTrigger& trigger_conf);
+  void RestoreMessage(const RecordMessage& msg);
+  const std::string GetDefaultOutputFile() const;
```

Comment:
const &.

Reply:
it said returning a reference to local variable:)

Reply:
It's not a reference, it's a copy.
1. If user want a local copy and modify later: `std::string file = GetDefaultOutputFile()`
2. If user just want a read only reference: `const std::string& file = GetDefaultOutputFile()`

So it's still better to return a const reference which saves memory for case 2.

And there is no reason to return a CONST copy. As for case 1 it will create two objects, the first one is const and returns by the function, the second one is non-const which is copy-constructed from the const one.

Reply:
got it, the warning with & was "returning reference to temporary", sorry for misleading:), so removed both const and reference like #1 suggested.